### PR TITLE
sc_sock_poll thread safety

### LINKED ISSUE #98: sc_sock thread-safety
I'm going to have a multi-threaded system with multiple poll objects and poller threads. I want to be able to add/del sockets to a poll object from multiple threads (e.g. server accepts a non-blocking socket in one thread and dispatches it to a random poller thread, or a worker thread does non-blocking socket connect and dispatches the socket to a random poller thread as well). 

As far as I can see `epoll` and `kqueue` already support this out of the box. Still `sc_sock_poll` has some additional unprotected shared state:
1. Resizeable  `sc_sock_poll.events` (and `count` + `cap`). This is fixable by removing resizing and having `events` buffer of a constant size (like 4096 or so).
2. Shared `sc_sock_poll.err`. Looks like the right way to handle it is to fill `sc_sock.err` instead of `sc_sock_poll.err` on add/del failures, but this is a backward compatibility breaking change (in a minor way though, it breaks only error handling path and not critically: the outdated code will see empty error message in `sc_sock_poll_err()` after failed add/del operations, which does not look too bad).

Windows as usual would require many more situps but still must be doable :)

@tezc What do you think about it? I can implement this, are you OK with reviewing/merging such a change?

### PR DESCRIPTION
**Pull request by svladykin**
Fixes https://github.com/tezc/sc/issues/98

For all implementations:
- replaced `err` with a static thread-local `sc_sock_poll_errstr` to make sure `sc_sock_poll_err()` will return correct error message for different threads;
- allocate constant size result buffer for `SC_SOCK_POLL_MAX_EVENTS` events, which is `1024` by default.

For Linux/FreeBSD/MacOS (`epoll` and `kqueue`) :
-  removed `count` and `cap` as well as `events` expanding to allow safe concurrent access (`epoll` and `kqueue` themselves are thread-safe); 
- all the updates to `sc_sock_fd` are done before `epoll_ctl`/`kevent` call to avoid possible data races, if the kernel call failed rollback to the original state;
- made the code as similar as possible between the two implementations to simplify further development;

For Windows (`WSAPoll`):
- added `lock` to protect shared state;
- added a separate `results` buffer which is filled under the lock with `sc_sock_poll_fill_results()` to allow `sc_sock_poll_event()` and `sc_sock_poll_data()` access results outside of the `lock`;
- now the resulting number of events returned by `sc_sock_poll_wait()` is the actual number of events collected, not the capacity of `events`;
- added `sc_sock_poll_data` struct which contains `index` of `fd` in `events`, `data` pointer and also volatile `edge_mask` which is updated by atomic instructions to allow Edge-Triggered mode work without acquiring the `lock`;
- added `data` as a data structure that does not use `realloc` to expand: this is needed to have a stable pointer from `sc_sock_fd` to `sc_sock_poll_data` (again no need to acquire `lock` on every access); 
- added `ops` buffer to store the socket operations submitted by other threads while polling is in progress;
- added `polling` flag that marks "polling is in progress" state and protects `events` from concurrent modifications;
- added `wakeup_pipe` which is registered to the same `sc_sock_poll` to be able to wake up the poller thread when submitting socket operations from parallel threads;
- if multiple operations are submitted by by parallel threads only the first one wakes up the poller thread, others do not need to do that;
- if multiple operations for the same socket were submitted before poller thread started processing them, they will be merged into a single operation;
-  if a submitted socket operation aimed to completely delete the socket from `sc_sock_poll` then no reference to the socket fdt is held in `ops`, so it is safe to do `sc_sock_term()` and `free()` on the socket right after successful call to `sc_sock_poll_del()` in any thread without waiting until the poller thread actually processes that operation.
- fixed error handling for `WSAPoll`: use `WSAGetLastError()` instead of `errno`;
- fixed error handling for `sc_sock_pipe_read()` and `sc_sock_pipe_write()`: no need to check `rc != len`;
- fixed bug in `sc_sock_pipe_init()`: need to assign read fd to fdt;

### CODE REVIEW

**Comment by svladykin** [comment]
@tezc Please review. 

As expected epoll/kqueue implementations got simplified (no `poll->events` resizing), windows got complicated (but I tried to comment extensively).

There are no backward incompatible changes from the API point of view: `poll->err` was replaced with a static thread-local, so now it is thread safe without any change in semantics. Maybe it is worth using this approach for sockets as well to avoid having 128 bytes long `err` in every socket.

I understand the PR is large, I can try to split it into multiple smaller ones, but the biggest part there is Windows and it would be hard to split.

**Comment by tezc** [comment]
@svladykin Thanks, I'll have a look in a couple of days.

**Comment by svladykin** [comment]
@tezc Did you have a chance to take a look?

**Comment by tezc** [comment]
@svladykin Sorry for the delay. I was a bit busy these days. I had a quick look and pushed some minor changes mostly about style. I need some more time to wrap my head around, hopefully I'll complete the review by tomorrow.

**Comment by svladykin** [comment]
@tezc No rush, I just wanted to make sure you did not forget about this PR :)

File: socket/sc_sock.c [hunk 1]
```diff
@@ -1424,38 +1399,45 @@ int sc_sock_poll_del(struct sc_sock_poll *p, struct sc_sock_fd *fdt,
...
+	// All the updates to fdt must be done before kevent() call to avoid
+	// data race: fdt can be published as *data here and this way a poller
+	// thread can see it partially updated if we do not follow that rule.
+	fdt->op = new_mask;
+
 	rc = kevent(p->fds, ev, count, NULL, 0, NULL);
+
 	if (rc != 0) {
+		// Rollback to the original state if failed.
+    	fdt->op = old_mask;
```
Comment (tezc) [inline socket/sc_sock.c]:
```suggestion
		// Rollback to the original state if failed.
		fdt->op = old_mask;
```

File: socket/sc_sock.c [hunk 2]
```diff
@@ -972,7 +978,7 @@ int sc_sock_pipe_read(struct sc_sock_pipe *p, void *data, unsigned int len)
 	int rc;
 
 	rc = recv(p->fds[0], (char *) data, len, 0);
-	if (rc == SOCKET_ERROR || (unsigned int) rc != len) {
+	if (rc == SOCKET_ERROR) {
```
Comment (tezc) [inline socket/sc_sock.c:981]: I don't have Windows machine atm but I guess `SOCKET_ERROR` is `-1`. Just wondering if we really return negative value on `SOCKET_ERROR`.  Windows return codes are mess, some error codes are zero or positive :)

File: socket/sc_sock.c [hunk 3]
```diff
@@ -1724,21 +1941,140 @@ uint32_t sc_sock_poll_event(struct sc_sock_poll *p, int i)
...
+			// wakeup_pipe, we just pick reasonably large buffer
+			// size for draining and assume that sc_sock_pipe_read()
+			// supports partial reads
+			// (less than the provided buffer size).
+			char buf[16];
+			sc_sock_pipe_read(&p->wakeup_pipe, buf, sizeof(buf));
+		}
+	} else {
+		// Because otherwise we would not even start polling.
+		assert(p->results_remaining == 0);
```
Comment (tezc) [inline socket/sc_sock.c:2061]: I think we'll hit here if `n == SOCKET_ERROR`. Should we move `n == SOCKET_ERROR` if check above?

File: socket/sc_sock.c [hunk 4]
```diff
@@ -1724,21 +1941,140 @@ uint32_t sc_sock_poll_event(struct sc_sock_poll *p, int i)
...
+			// of p->ops, this way we do not drain any extra wakeup
+			// signals.
+			//
+			// At this point we may have any number of bytes in
+			// wakeup_pipe, we just pick reasonably large buffer
+			// size for draining and assume that sc_sock_pipe_read()
+			// supports partial reads
+			// (less than the provided buffer size).
+			char buf[16];
+			sc_sock_pipe_read(&p->wakeup_pipe, buf, sizeof(buf));
```
Comment (tezc) [inline socket/sc_sock.c]: Maybe we can assert return value is not negative.

File: socket/sc_sock.c [hunk 5]
```diff
@@ -1538,8 +1549,18 @@ int sc_sock_poll_term(struct sc_sock_poll *p)
 		return 0;
 	}
 
+	sc_sock_poll_del(p, &p->wakeup_pipe.fdt, SC_SOCK_READ, NULL);
+	sc_sock_pipe_term(&p->wakeup_pipe);
```
Comment (tezc) [inline socket/sc_sock.c]: maybe we can check error here and then call `sc_sock_poll_set_err()` on error

**Comment by tezc** [comment]
@svladykin Sorry for the delay. I didn't have much time to review and probably it will take more time to wrap my head around the changes fully. Not to make this review longer, considering API is compatible and most changes are on Windows, I think we can merge it and I'll take a look in parallel and ask you questions if I need :)   

- Thanks for the pipe & poll resizing fixes.
- Regarding using `__thread` for sock err, I didn't want to use anything compiler specific initially but socket code won't be quite portable anyway (at least not as much as a linkedlist impl). So, I guess we can use `__thread` for sockets as well. 
- Overall, Windows side changes make sense but still I need more time to understand it fully. I can do that in parallel after we merge this anyway. I'll catch up in a few weeks hopefully. 

Added a few comments in the PR regarding return value checks, after those, we can merge the PR I guess. 
Thanks for the PR and sorry for late review. Last month was a bit busy and looks like there is another busy month upcoming for me :)

File: socket/sc_sock.c (hunk 2 above)
Comment (svladykin) [reply socket/sc_sock.c:981]: Yes, it is `-1` : https://learn.microsoft.com/en-us/windows/win32/winsock/handling-winsock-errors

File: socket/sc_sock.c (hunk 3 above)
Comment (svladykin) [reply socket/sc_sock.c:2061]: We will also hit here if either `n == SOCKET_ERROR` or `n == 0`, in both cases we should expect `results_remaining` to be `0`. I will add assert and comment for that.

File: socket/sc_sock.c (hunk 4 above)
Comment (svladykin) [reply socket/sc_sock.c]: Good point, will add it.

File: socket/sc_sock.c (hunk 5 above)
Comment (svladykin) [reply socket/sc_sock.c]: Makes sense.

**Comment by svladykin** [comment]
@tezc  Thanks for the review. I will address your comments shortly.

**Comment by svladykin** [comment]
@tezc I have made the requested adjustments. Also I fixed a `WSAPoll` error handling issue on Windows. Please look.

File: socket/sc_sock.c [hunk 6]
```diff
@@ -1588,22 +1650,152 @@ static int sc_sock_poll_expand(struct sc_sock_poll *p)
...
+	}
+
+	o = &ops[index];
+
+	if (add) {
+		o->add_events |= events;
+		o->del_events &= ~events;
+	} else {
+		o->add_events &= ~events;
+        o->del_events |= events;
```
Comment (tezc) [inline socket/sc_sock.c]:
```suggestion
		o->add_events &= ~events;
		o->del_events |= events;
```

**Comment by tezc** [comment]
@svladykin Thank you very much, looks good to me. We have %100 branch coverage on Linux but other OSes like Windows, damn there are many bugs. Thank you for the fixes. 

Is there anything else we need to do for this PR?  Btw could you add a few lines to the top comment and briefly describe the changes?

**Comment by svladykin** [comment]
@tezc I believe this PR is good to go. The `epoll`/`kqueue` changes seems to be safe, Windows is less safe, but if any issues arise I will fix them.

**Review by tezc — APPROVED** [review]

**Pull request merged**